src/jit/lower.cpp

```diff
@@ -2428,12 +2430,27 @@ void Lowering::LowerCompare(GenTree* cmp)
                 // situations. In particular this include CALL, sometimes the JIT unnecessarilly widens
                 // the result of bool returning calls.
                 //
+                bool removeCast =
+#ifdef _TARGET_ARM64_
+                    (op2Value <= ((castToType == TYP_BOOL) ? 1 : 0)) && cmp->OperIs(GT_EQ, GT_NE, GT_GT) &&
+#endif
+                    (castOp->OperIs(GT_CALL, GT_LCL_VAR) || castOp->OperIsLogical() || IsContainableMemoryOp(castOp));
 
-                if (castOp->OperIs(GT_CALL, GT_LCL_VAR) || castOp->OperIsLogical() || IsContainableMemoryOp(castOp))
+                if (removeCast)
                 {
                     assert(!castOp->gtOverflowEx()); // Must not be an overflow checking operation
 
+#ifdef _TARGET_ARM64_
+                    bool cmpEq = cmp->OperIs(GT_EQ);
+
+                    if (op2Value != 0)
+                        cmpEq = !cmpEq;
+
+                    cmp->SetOperRaw(cmpEq ? GT_TEST_EQ : GT_TEST_NE);
+                    op2->SetIconValue((op2Value == 0) ? 0xff : 0x1);
```

Comment:
If C# is supposed to CAST_BOOL is supposed to behave the same way as C/C++, then this is should be

```
                    op2->SetIconValue((castToType == TYP_BOOL) ? (~0UL >> 8* (sizeof(uint64_t) - castOpSize)) : 0xff);
```

Reply:
Since C# IL does not behave the same way as C++/C, The above comment can be ignored